bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -0,0 +1,171 @@
+package com.bugsnag.android.internal
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import java.io.File
+import java.io.IOException
+import java.lang.IllegalStateException
+import java.nio.BufferOverflowException
+import java.util.function.BiConsumer
+
+/**
+ * A document with journal backing.
+ * Document snapshots are persisted to files, and journal entries are persisted to a memory-mapped
+ * file.
+ *
+ * Changes to the document must me made only via journal commands. Do not modify the document or any
+ * of its sub-components directly!
+ */
+class JournaledDocument
+/**
+ * Constructor.
+ * Upon construction, this document will have associated snapshot and journal files.
+ *
+ * @param baseDocumentPath The base path + filename to derive snapshot and journal names from.
+ * @param journalType A string identifying the type of journal (used for verification when the
+ *                    journal is reloaded)
+ * @param version The journal type version (to support migration)
+ * @param bufferSize The size of the shared memory buffer (and thus the maximum size of the journal)
+ * @param initialDocument The initial document contents (will be shallow copied)
+ */
+constructor(
+    baseDocumentPath: File,
+    journalType: String,
+    version: Int,
+    bufferSize: Long,
+    initialDocument: MutableMap<String, Any>
+) : Map<String, Any> {
+    private val journalPath = File(baseDocumentPath.path + journalExtension)
+    private val snapshotPath = File(baseDocumentPath.path + snapshotExtension)
+    private val newSnapshotPath = File(baseDocumentPath.path + newSnapshotExtension)
+
+    private val document = initialDocument.toMutableMap()
+    private val journal = Journal(journalType, version)
+    private val journalStream = MemoryMappedOutputStream(journalPath, bufferSize, clearedByteValue)
+    private var isOpen = true
+
+    init {
+        // Make sure any leftover files from the last instance at this path are gone.
+        newSnapshotPath.delete()
+        snapshot()
+    }
+
+    val journalType: String get() = this.journal.type
+    val version: Int get() = this.journal.version
+
+    /**
+     * Add a journal command.
+     *
+     * This will first serialize the command to shared memory. If the memory is too full,
+     * it will create a snapshot to free up some room and try a second time.
+     * Next, it will apply the command to the document, and then add it to the journal itself.
+     */
+    fun addCommand(command: Journal.Command) {
+        if (!isOpen) throw IllegalStateException("Cannot add commands to a closed document")
+        try {
+            command.serialize(journalStream)
+        } catch (ex: BufferOverflowException) {
+            snapshot()
+            command.serialize(journalStream)
```

Comment:
This could (theoretically) fail with the same exception which would then be uncaught and terminate the process. Although it's unlikely for common cases it'd be good to handle this scenario.

The `serialize` method also throws an `IOException` - should that be handled here or higher up?

Reply:
If it throws again, there's a serious problem. Any exceptions that escape this function must be handled by the caller (which will be the wrapping journal API that handles errors, enforces thread safety, etc).